Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py

```diff
@@ -0,0 +1,714 @@
+        if this_IoverSig > prev_IoverSig:
+            prev_IoverSig = this_IoverSig
+            nbad = 0
+        else:
+            nbad += 1
+            if nbad > ntol:
+                break
+    istep_hi -= nbad
+    nbad = 0
+    prev_IoverSig = calc_snr(signal[ilo:ihi], error_sq[ilo:ihi])
+    for istep_lo in range(1, ilo):
+        this_IoverSig = calc_snr(signal[ilo - istep_lo:ihi], error_sq[ilo - istep_lo:ihi])
+        if this_IoverSig > prev_IoverSig:
+            prev_IoverSig = this_IoverSig
+            nbad = 0
+        else:
+            nbad += 1
+            if nbad > ntol:
+                break
+    istep_lo -= nbad
+    return ilo - istep_lo, ihi + istep_hi
+
+
+def find_peak_limits(signal, error_sq, ilo, ihi):
+    # find initial background points in tof window using skew method (excl. points above mean)
+    ibg, _ = find_bg_pts_seed_skew(signal[ilo:ihi])
+    # create mask and find largest contiguous region of peak bins
+    skew_mask = np.ones(ihi - ilo, dtype=bool)
+    skew_mask[ibg] = False
+    labels, nlabel = label(skew_mask)
+    if nlabel == 0:
+        ilabel = 0
+    else:
+        ilabel = np.argmax([np.sum(labels == ilabel) for ilabel in range(1, nlabel + 1)]) + 1
+    istart, iend = np.flatnonzero(labels == ilabel)[[0, -1]] + ilo
+    # expand window to maximise I/sig (good for when peak not entirely in window)
+    return optimise_window_intens_over_sig(signal, error_sq, istart, iend + 1)  #
+
+
+def focus_data_in_detector_mask(signal, error, peak_mask, non_bg_mask, ixpk):
+    # focus peak in TOF
+    ypk = signal[peak_mask].sum(axis=0)
+    epk_sq = np.sum(error[peak_mask] ** 2, axis=0)
+    # get background shell of non peak/feature pixels
+    kernel = np.ones((3, 3))
+    bg_shell_mask = convolve2d(peak_mask, kernel, mode='same')
+    norm = convolve2d(np.ones(bg_shell_mask.shape), kernel, mode='same')
+    bg_shell_mask = (bg_shell_mask / norm) > 0
+    bg_shell_mask = np.logical_and(bg_shell_mask, ~non_bg_mask)
+    # focus background shell
+    scale = peak_mask.sum() / bg_shell_mask.sum()
+    ybg = scale * signal[bg_shell_mask].sum(axis=0)
+    ebg_sq = (scale ** 2) * np.sum(error[bg_shell_mask] ** 2, axis=0)
+    # replace zero errors in epk and ebg_sq  with avg. error in same quarter of spectrum in background data
+    width = len(ypk) / 4
+    iquarter = ixpk // width
+    ebg_sq_subset = ebg_sq[int(iquarter * width):int((iquarter + 1) * width)]
+    ebg_sq_avg = np.mean(ebg_sq_subset[ebg_sq_subset > 0])
+    epk_sq[epk_sq == 0] = ebg_sq_avg
+    ebg_sq[ebg_sq == 0] = ebg_sq_avg
+    # subtract bg from focused peak
+    ypk = ypk - ybg  # removes background and powder lines (roughly line up in tof)
+    epk_sq = epk_sq + ebg_sq
+    return ypk, epk_sq
+
+
+def find_peak_mask(signal, ixlo, ixhi, irow, icol, use_nearest):
+    _, ipeak2D = find_bg_pts_seed_skew(signal[:, :, ixlo:ixhi].sum(axis=2).flatten())
```

Comment:
Please can you add a comment indicating what dimension you're calculating the skew over here. I think you're calculating it over detectors (which is different to the other call to `find_bg_pts_seed_skew` where it's being calculated over TOF bins). It's not always totally obvious what the 3 dimensions of the signal array are so the comment would help clarify this

Reply:
The function `find_bg_pts_seed_skew` works on any 1D array - it is used both to search for background points in the tof window and the detector window (the latter array is flattened). In this case it is used in the detector window (after summming over a TOF range). I will add a comment and maybe change the function name to `find_2D_peak_mask` or similar?

Reply:
OK great. I think the function name is OK (was clear to me what the peak mask was). Just lost my bearings for a while about what was going on in `find_bg_pts_seed_skew` until I realised it was general and worked on any 1D array